### PR TITLE
[metabase-lib] Port SaveQuestionModal, QueryModals, QB.View

### PR DESCRIPTION
**Pull request by bshepherdson**
Now they use Question rather than Card.

### CODE REVIEW

File: frontend/src/metabase/query_builder/components/QueryModals.jsx [hunk 1]
```diff
@@ -75,17 +75,17 @@ class QueryModals extends React.Component {
...
+            await this.props.onSave(question);
             onCloseModal();
           }}
-          onCreate={async card => {
-            await this.props.onCreate(card);
+          onCreate={async question => {
+            await this.props.onCreate(question);
             if (question.isDataset()) {
               onCloseModal();
               setQueryBuilderMode("view");
```
Comment (bshepherdson) [inline frontend/src/metabase/query_builder/components/QueryModals.jsx:91]: @ranquild This is where the timing problem is - the new `onCreate` is apparently a frame or two faster, and so it leaves the editor too fast for the test.

**Comment by camsaul** [comment]
Did you mean to commit that change to the sample dataset file?

**Review by camsaul — APPROVED** [review]
backend parts look good to me 😆 

I looked thru the FE code too and as far as I can tell it looks good too.
I'm hoping at some point we can start using MLv2 to generate the queries we use in new tests rather than hardcode MBQL since at some point we'll want to eliminate all of the hardcoded MBQL in JS tests

File: frontend/src/metabase-lib/Question.ts [hunk 2]
```diff
@@ -1286,6 +1298,211 @@ class QuestionInner {
     return hasQueryBeenAltered ? question.markDirty() : question;
   }
 
+  generateQueryDescription(tableMetadata, options = {}) {
```
Comment (bshepherdson) [inline frontend/src/metabase-lib/Question.ts:1301]: This logic is from `queries/utils/description.js`. It'll probably be moved to CLJC functions soon; it makes a nice proof of concept. (Lots of logic, simple return value, self-contained.)

File: frontend/test/metabase/containers/SaveQuestionModal.unit.spec.js [hunk 3]
```diff
@@ -90,12 +90,9 @@ function getDirtyQuestion(originalQuestion) {
...
     .question();
-  return question.setCard({
-    ...question.card(),
-    // After a saved question is edited, the ID gets removed
-    // and a user can either overwrite a question or save it as a new one
-    id: undefined,
-  });
+  // After a saved question is edited, the ID gets removed
+  // and a user can either overwrite a question or save it as a new one
+  return question.setId(undefined);
```
Comment (ranquild) [inline frontend/test/metabase/containers/SaveQuestionModal.unit.spec.js:95]: should it be this? https://github.com/metabase/metabase/blob/bc782f534a72de567b885bb921e4f9fb816c36a9/frontend/src/metabase-lib/Question.ts#L160

**Comment by kulyk** [comment]
@bshepherdson I see `resources/sample-database.db.mv.db` in the diff, is that intentional?

File: frontend/test/metabase/containers/SaveQuestionModal.unit.spec.js [hunk 4]
```diff
@@ -187,8 +217,8 @@ describe("SaveQuestionModal", () => {
       await waitFor(() => {
         expect(onCreateMock).toHaveBeenCalledTimes(1);
       });
-      expect(onCreateMock).toHaveBeenCalledWith({
-        ...question.card(),
+      checkCall(onCreateMock, {
```
Comment (ranquild) [inline frontend/test/metabase/containers/SaveQuestionModal.unit.spec.js:220]: I find this new abstraction hard to follow, and it would be great to simplify this. E.g.
```
const newQuestion = onCreateMock.mock.calls[0];
expect(newQuestion.displayName()).toEqual(EXPECTED_SUGGESTED_NAME);
expect(newQuestion.description()).toBeNull();
expect(newQuestion.collectionId()).toBeNull();
```

**Comment by bshepherdson** [comment]
> @bshepherdson I see `resources/sample-database.db.mv.db` in the diff, is that intentional?

Ah, good catch. That crept back in, there's an issue where running Cypress tests locally can change your sample DB. I'll drop it from the diff before submitting.

**Review by ranquild — COMMENTED** [review]
Please take a look at the comments

File: frontend/src/metabase-lib/Question.ts [hunk 5]
```diff
@@ -1286,6 +1298,211 @@ class QuestionInner {
...
+    const sections = options.sections
+      .map(section =>
+        _.flatten(sectionFns[section](tableMetadata, query, options)).filter(
+          s => !!s,
+        ),
+      )
+      .filter(s => s && s.length > 0);
+
+    const description = _.flatten(DESCRIPTION.joinList(sections, ", "));
+    return description.join("");
```
Comment (kulyk) [inline frontend/src/metabase-lib/Question.ts:1338]: FWIW, there is an `Array.prototype.flat` method we can use instead of `_.flatten`

File: frontend/src/metabase-lib/Question.ts [hunk 6]
```diff
@@ -1286,6 +1298,211 @@ class QuestionInner {
...
+
+    // these array gymnastics are needed to support JSX formatting
+    const query = this.datasetQuery().query;
+    const sections = options.sections
+      .map(section =>
+        _.flatten(sectionFns[section](tableMetadata, query, options)).filter(
+          s => !!s,
+        ),
+      )
+      .filter(s => s && s.length > 0);
```
Comment (kulyk) [inline frontend/src/metabase-lib/Question.ts:1335]: Can we avoid single char variables here?

File: frontend/src/metabase-lib/Question.ts [hunk 7]
```diff
@@ -1286,6 +1298,211 @@ class QuestionInner {
...
+      limit: this._getLimitDescription.bind(this),
+    };
+
+    // these array gymnastics are needed to support JSX formatting
+    const query = this.datasetQuery().query;
+    const sections = options.sections
+      .map(section =>
+        _.flatten(sectionFns[section](tableMetadata, query, options)).filter(
+          s => !!s,
+        ),
```
[the remark this answers is not in the document]
Comment (kulyk) [reply frontend/src/metabase-lib/Question.ts:1333]: Btw, I believe we prefer `list.filter(Boolean)` over `list.filter(element => !!element)` on the FE

File: frontend/src/metabase-lib/Question.ts [hunk 8]
```diff
@@ -1286,6 +1298,211 @@ class QuestionInner {
...
+          case "min":
+            return [
+              t`Minimum of `,
+              this._getFieldName(tableMetadata, aggregation[1], options),
+            ];
+          default:
+            return [formatExpression(aggregation, { tableMetadata })];
+        }
+      }),
+      "and",
```
Comment (kulyk) [inline frontend/src/metabase-lib/Question.ts:1436]: Should we translate this?

File: frontend/src/metabase-lib/Question.ts [hunk 9]
```diff
@@ -1286,6 +1298,211 @@ class QuestionInner {
...
+    );
+  }
+
+  private _getBreakoutDescription(tableMetadata, { breakout }, options) {
+    if (breakout && breakout.length > 0) {
+      return [
+        t`Grouped by `,
+        DESCRIPTION.joinList(
+          breakout.map(b => this._getFieldName(tableMetadata, b, options)),
+          " and ",
```
Comment (kulyk) [inline frontend/src/metabase-lib/Question.ts:1446]: Should we translate this?

File: frontend/src/metabase-lib/Question.ts [hunk 10]
```diff
@@ -1286,6 +1298,211 @@ class QuestionInner {
...
+        ),
+      ];
+    }
+  }
+
+  private _getFilterDescription(tableMetadata, query, options) {
+    // getFilters returns list of filters without the implied "and"
+    const filters = ["and"].concat(QUERY.getFilters(query));
+    if (filters && filters.length > 1) {
```
Comment (kulyk) [inline frontend/src/metabase-lib/Question.ts:1455]: Could be `filters?.length > 1`

File: frontend/src/metabase-lib/Question.ts [hunk 11]
```diff
@@ -1286,6 +1298,211 @@ class QuestionInner {
...
+    if (filter[0] === "and" || filter[0] === "or") {
+      const clauses = filter
+        .slice(1)
+        .map(f => this._getFilterClauseDescription(tableMetadata, f, options));
+      return DESCRIPTION.conjunctList(clauses, filter[0].toLowerCase());
+    } else if (filter[0] === "segment") {
+      const segment = _.findWhere(tableMetadata.segments, { id: filter[1] });
+      return segment ? segment.name : "[Unknown Segment]";
+    } else if (filter[0] === "between" && filter[1][0] === "+") {
+      return this._getFieldName(tableMetadata, filter[1][1], options);
```
Comment (kulyk) [inline frontend/src/metabase-lib/Question.ts:1473]: Would it make sense to give names to things like `filter[1][1]`? Could make it a bit easier to follow, otherwise, people would need to recall how MBQL looks like every time

File: frontend/src/metabase-lib/Question.ts [hunk 12]
```diff
@@ -1286,6 +1298,211 @@ class QuestionInner {
...
+          orderBy.map(([direction, field]) => {
+            const name = FIELD_REF.isAggregateField(field)
+              ? this._getAggregationDescription(tableMetadata, query, options)
+              : this._getFieldName(tableMetadata, field, options);
+
+            return (
+              name + " " + (direction === "asc" ? "ascending" : "descending")
+            );
+          }),
+          " and ",
```
Comment (kulyk) [inline frontend/src/metabase-lib/Question.ts:1494]: Should we translate this?

File: frontend/test/metabase/containers/SaveQuestionModal.unit.spec.js [hunk 13]
```diff
@@ -111,6 +108,39 @@ function fillForm({ name, description }) {
   }
 }
 
+function checkCall(
+  mockFn,
+  { question, name, id, description, collection_id } = {},
+) {
+  const call = mockFn.mock.calls[0];
+  expect(call.length).toBe(1);
```
Comment (kulyk) [inline frontend/test/metabase/containers/SaveQuestionModal.unit.spec.js:116]: `expect(mockFn).toHaveBeenCalledTimes(1)`

File: frontend/test/metabase/containers/SaveQuestionModal.unit.spec.js [hunk 14]
```diff
@@ -111,6 +108,39 @@ function fillForm({ name, description }) {
...
+  }
+
+  // Check a few specific properties first for clearer debugging.
+  expect(call[0].id()).toBe(newQuestion.id());
+  expect(call[0].displayName()).toBe(newQuestion.displayName());
+  expect(call[0].description()).toBe(newQuestion.description());
+  expect(call[0].collectionId()).toEqual(newQuestion.collectionId());
+
+  // Then checking the overall isEqual() for completeness.
+  expect(call[0].isEqual(newQuestion)).toBe(true);
```
Comment (kulyk) [inline frontend/test/metabase/containers/SaveQuestionModal.unit.spec.js:141]: nit: maybe assign `call[0]` to a variable to make it clear it's a question?

```js
const [question] = call
expect(question.id()).toBe(newQuestion.id())
```

**Review by kulyk — COMMENTED** [review]
Awesome work! 🏆 Left a few questions and ideas

File: frontend/test/metabase/containers/SaveQuestionModal.unit.spec.js (hunk 4 above)
Comment (kulyk) [reply frontend/test/metabase/containers/SaveQuestionModal.unit.spec.js:220]: Yeah, we don't add functions for asserting on stuff. Eslint would also complaint about not having any assertions in a test if there's no `expect` in the test body and only `checkCall`

File: frontend/src/metabase-lib/Question.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (bshepherdson) [reply frontend/src/metabase-lib/Question.ts:1333]: As discussed on Slack: I'm planning to port (and refactor) this whole block to CLJC starting today or Monday, so we're agreed it's not worth investing in improving this code here.

File: frontend/src/metabase-lib/Question.ts (hunk 8 above)
Comment (bshepherdson) [reply frontend/src/metabase-lib/Question.ts:1436]: I've added TODOs to the various untranslated strings. I'll try to find a more i18n-friendly way to do these text lists. Not all languages structure lists the same way, of course. Even the above translated strings `"Maximum of "` etc. are naively concatenated rather than using i18n placeholders, so they assume a certain syntax. We can do better!

File: frontend/test/metabase/containers/SaveQuestionModal.unit.spec.js (hunk 3 above)
Comment (bshepherdson) [reply frontend/test/metabase/containers/SaveQuestionModal.unit.spec.js:95]: Right idea, but the wrong method - the name is supposed to be preserved here. We actually want `.markDirty()`, which it now uses.

File: frontend/test/metabase/containers/SaveQuestionModal.unit.spec.js (hunk 13 above)
Comment (bshepherdson) [reply frontend/test/metabase/containers/SaveQuestionModal.unit.spec.js:116]: That's how it's done now.

File: frontend/test/metabase/containers/SaveQuestionModal.unit.spec.js (hunk 14 above)
Comment (bshepherdson) [reply frontend/test/metabase/containers/SaveQuestionModal.unit.spec.js:141]: Done.

File: frontend/test/metabase/containers/SaveQuestionModal.unit.spec.js (hunk 4 above)
Comment (bshepherdson) [reply frontend/test/metabase/containers/SaveQuestionModal.unit.spec.js:220]: I've factored this back into each test.

**Comment by bshepherdson** [comment]
Comments addressed; PTAL.

Note the punt on improving the `generateQueryDescription` code I moved to `Question` - it'll be ported to CLJC shortly.

**Review by ranquild — APPROVED** [review]

**Review by kulyk — APPROVED** [review]
